tools/slidetool.1.in
.\"
.\" OpenSlide, a library for reading whole slide image files
.\"
.\" Copyright (c) 2007-2012 Carnegie Mellon University
.\" Copyright (c) 2023      Benjamin Gilbert
.\" All rights reserved.
.\"
.\" OpenSlide is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU Lesser General Public License as
.\" published by the Free Software Foundation, version 2.1.
.\"
.\" OpenSlide is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU Lesser General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with OpenSlide. If not, see
.\" <http://www.gnu.org/licenses/>.
.\"


.\" See man-pages(7) for formatting conventions.


.TH SLIDETOOL 1 2023-07-27 "OpenSlide @SUFFIXED_VERSION@" "User Commands"

.mso www.tmac

.SH NAME
slidetool \- Retrieve data from whole slide images

.SH SYNOPSIS
.BR "slidetool " [ \-\-help "] [" \-\-version ]
.br
.B slidetool assoc icc read
.IR "file name" " [" output-file ]
.br
.B slidetool assoc list
.IR file ...
.br
.B slidetool assoc read
.IR "file name" " [" output-file ]
.br
.B slidetool prop get
.IR "property file" ...
.br
.BR "slidetool prop list" " [" \-\-names ]
.IR file ...
.br
.B slidetool region icc read
.IR file " [" output-file ]
.br
.B slidetool region read
.IR "file x y level width height" " [" output-file ]
.br
.B slidetool slide open
.IR file ...
.br
.B slidetool slide quickhash1
.IR file ...
.br
.B slidetool slide vendor
.IR file ...

.SH DESCRIPTION
.B slidetool
uses OpenSlide to retrieve data from whole slide images,
also known as virtual slide files.

.SS slidetool assoc icc read
Write the associated image's ICC color profile to
.I output-file
if specified, and otherwise to standard output.
If no ICC color profile is available, fail.

.SS slidetool assoc list
List associated images for one or more slides.
Associated images are small non-pyramidal images containing
additional information about the slide, such as its label.

.SS slidetool assoc read
Write the associated image with the specified
.I name
to a PNG image.
If
.I output-file
is not specified, the image will be written to standard output.

.SS slidetool prop get
Print a single OpenSlide property value for one or more slides.
Properties are individual pieces of textual metadata about the slide.
Properties prefixed with
.I openslide.
are defined by OpenSlide;
others are defined by the individual slide format.

.SS slidetool prop list
Print all OpenSlide properties for one or more slides.

.SS slidetool region icc read
Write the slide's ICC color profile to
.I output-file
if specified, and otherwise to standard output.
If no ICC color profile is available, fail.

.SS slidetool region read
Write a region of the specified
.I level
of a slide to a PNG image.
.I x
and
.I y
are pixel coordinates in level 0, the highest-resolution slide level.
.I width
and
.I height
are pixel dimensions in the specified
.IR level .
If
.I output-file
is not specified, the image will be written to standard output.

The dimensions of each level of a slide can be obtained with
.BR "slidetool prop list" .

.SS slidetool slide open
Check whether OpenSlide can open one or more slide files.

.SS slidetool slide quickhash1
Print OpenSlide
.I quickhash-1
checksums for one or more slides, in a format similar to
.BR sha256sum (1).
.PP
.I quickhash-1
is a non-cryptographic, 256-bit hash of a subset of a slide's data.
It uniquely identifies a particular slide,
but cannot be used to detect corruption or modification of the slide file.
.PP
.I quickhash-1
is not defined for all slide files supported by OpenSlide.
If a slide does not have a
.IR quickhash-1 ,
"No quickhash\-1 available" will be printed.

.SS slidetool slide vendor
Report the detected OpenSlide vendor name for one or more slide files.

.SH OPTIONS
.TP
.B \-\-help
Display usage summary.

.TP
.B \-\-names
For
.BR "slidetool prop list" ,
omit property values.

.TP
.B \-\-version
Display version and copyright information.

.SH EXIT STATUS
.B slidetool
returns 0 on success, 1 if a slide file could not be read,
or 2 if the arguments are invalid.

.SH COPYRIGHT
Copyright \(co 2007-2026 Carnegie Mellon University and others
.PP
OpenSlide is free software: you can redistribute it and/or modify it
under the terms of the
.URL http://gnu.org/licenses/lgpl-2.1.html "GNU Lesser General Public License, version 2.1" .
.PP
OpenSlide comes with NO WARRANTY, to the extent permitted by law.
See the GNU Lesser General Public License for more details.

.SH SEE ALSO
.BR openslide\-quickhash1sum (1),
.BR openslide\-show\-properties (1),
.BR openslide\-write\-png (1),
.BR sha256sum (1)


CHANGELOG.md
# Notable Changes in OpenSlide

## Version 4.0.1, 2026-06-07

### Security

* Fix arbitrary memory write with crafted Ventana BIF file (CVE-2026-48977)
* Fix return of uninitialized pixels with libtiff 4.7.1 (CVE-2026-54604)

### Format support

* New formats: ARGOS AVS, Huron TIFF
* New format: Zeiss CZI, uncompressed or zstd-compressed (thanks, Wei Chen)
* aperio: Support label and macro images on newer slides
* aperio: Add main image ICC profile to thumbnail if indicated by metadata
* dicom: Support levels split into multiple files (thanks, John Cupitt)
* dicom: Support reversible JPEG 2000 compression
* dicom: Fix colors with RGB JPEG (thanks, John) and irreversible JPEG 2000
* dicom: Support files with Image Type `ORIGINAL\PRIMARY\VOLUME\RESAMPLED`
* dicom: Fix swapped X and Y MPP
* generic-tiff: Support TIFFs with missing tiles
* hamamatsu: Fully support NDPI larger than 4 GB (thanks, Nick Trahearn)
* hamamatsu: Fix errors reading highly downsampled NDPI levels
* hamamatsu: Support NDPI files without `RowsPerStrip`
* hamamatsu: Correctly read NDPI ASCII strings shorter than 4 characters
* mirax: Support missing or variant `OBJECTIVE_MAGNIFICATION`
* mirax: Ignore empty `Slidedat.ini` key names

### New features

* Add `openslide.barcode` property
* Support building directly on Windows with GCC or clang-cl
* Add `slidetool test deps` command to run self-tests

### Changes

* Require Zstandard
* Require Meson ≥ 0.55; libdicom ≥ 1.3; libjpeg-turbo ≥ 1.3 or libjpeg ≥ 9c
* Replace GDK-PixBuf dependency with built-in BMP parser
* Improve I/O performance with DICOM, Hamamatsu, TIFF
* Prevent libtiff ≥ 4.5.0 from logging to stderr

### Bug fixes

* Fix filename character set of command-line tool output files on Windows


## Version 4.0.0, 2023-10-11

### Breaking changes

* Update soname to `libopenslide.so.1`
* Remove all deprecated functions
* Clear `openslide_read_associated_image()` output buffer on error
* Remove undocumented NULL `dest` support in `openslide_read_associated_image()`
* Change GLib log domain to `OpenSlide`
* Convert build system to Meson (thanks, Jan Harkes)

### New features

* New format: DICOM WSI (thanks, John Cupitt and Jim O'Donnell)
* Add APIs to read ICC color profiles (thanks, John)
* Add APIs to configure tile cache size and share caches between slides
* Add properties for associated image metadata
* generic-tiff: Set MPP properties if available
* philips: Set objective power property if available
* Add `slidetool` command-line tool which supports all OpenSlide features
* Combine all command-line tools into the same binary
* Add ICC profile to PNG images written by command-line tools
* Add self-test that doesn't require sample data (run with `meson test`)

### Changes

* Require libtiff ≥ 4, OpenJPEG ≥ 2.1, GLib ≥ 2.56, SQLite ≥ 3.14
* Require libdicom ≥ 1.0, with build-time fallback for now
* Remove support for including `openslide.h` in Visual Studio \< 2013
* Fail `openslide_open()` with a broken pixman 0.38.x
* Avoid extra buffer copy in `openslide_read_region()`
* Replace `goto`-based cleanup with `g_autoptr`
* Stop using deprecated GLib slice allocator
* Use internal wrappers for file I/O
* Documentation improvements

### Bug fixes

* Use UTF-8 filenames on Windows
* Improve `openslide-write-png` performance for very large regions
* Fix assertions on JPEG decode errors when compiled with Clang
* Portability fixes (thanks, Billy Robert O'Neal III and Kleis Auke Wolthuizen)
* aperio: Set objective power property even if floating point
* hamamatsu: Fix `Restart marker not found` on VMS slides with multiple Z-layers
* hamamatsu: Fix integer overflow in VMS parsing (thanks, Adam Goode)
* mirax: Fix `Expected 1 value` error
* philips: Don't cache missing tiles
* sakura: Fix memory leak reading missing tile


## Version 3.4.1, 2015-04-20

* New formats: Philips TIFF, Ventana TIFF
* Support OpenJPEG 2.1.0
* Improve performance of JPEG and JP2K decoding
* Add `openslide.region[i].*` properties
* Improve MATLAB compatibility
* Enable function deprecation warnings with MSVC
* Many portability fixes
* aperio: Detect OpenJPEG chroma subsampling breakage during `open`
* aperio: Fill in missing tiles with downsampled data
* aperio: Report MPP for slides scanned in locales with decimal comma
* hamamatsu: Support NDPI files \> 4 GB
* hamamatsu: Properly detect NDPI slides produced by NDP.toolkit
* hamamatsu: Support VMS/VMU slides without a `NoLayers` key
* hamamatsu: Report MPP for VMS/VMU
* leica: Support slides with `2010/03/10` XML namespace
* leica: Base64-decode `leica.barcode` property in `2010/10/01` namespace
* sakura: Support slides with multiple focal planes
* sakura: Support slides without tile table
* ventana: Support slides with multiple focal planes
* ventana: Improve positioning of AOIs within level
* ventana: Fix failure to recognize macro image on some slides


## Version 3.4.0, 2014-01-25

* Major internal restructuring
* New formats: Hamamatsu NDPI, Sakura SVSLIDE, Ventana BIF (preliminary)
* Add `openslide_detect_vendor()`
* Deprecate `openslide_can_open()` (not very useful and often misused)
* Document performance considerations for `openslide_open()`
* Add properties giving the bounds of the non-empty region of the slide
* leica: Support multiple main images if their levels are coplanar
* leica: Use slide size as level size
* mirax: Support PNG- and BMP-formatted slides
* mirax: Fix `Expected first 0 value` error opening some slides
* mirax: Fix incorrect tile placement on some slides without overlaps
* mirax: Never synthesize downsampled levels
* Add `OPENSLIDE_DEBUG` environment variable (`OPENSLIDE_DEBUG=?` for help)
* Fix some crashes in error paths
* Add tests for many error paths


## Version 3.3.3, 2013-04-13

* Fix inclusion of `openslide.h` with MSVC
* Properly handle Aperio JP2K slides with zero-length tiles
* Support Hamamatsu slides with blank `MacroImage` key


## Version 3.3.2, 2012-12-01


tools/slidetool.c
/*
 *  OpenSlide, a library for reading whole slide image files
 *
 *  Copyright (c) 2007-2012 Carnegie Mellon University
 *  Copyright (c) 2015-2023 Benjamin Gilbert
 *  All rights reserved.
 *
 *  OpenSlide is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as
 *  published by the Free Software Foundation, version 2.1.
 *
 *  OpenSlide is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with OpenSlide. If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <string.h>
#include <glib.h>
#include "openslide.h"
#include "openslide-common.h"
#include "slidetool.h"
#include "config.h"

static const char *version_format = "%s " SUFFIXED_VERSION ", "
"using OpenSlide %s\n"
"Copyright (C) 2007-2026 Carnegie Mellon University and others\n"
"\n"
"OpenSlide is free software: you can redistribute it and/or modify it under\n"
"the terms of the GNU Lesser General Public License, version 2.1.\n"
"<http://gnu.org/licenses/lgpl-2.1.html>\n"
"\n"
"OpenSlide comes with NO WARRANTY, to the extent permitted by law.  See the\n"
"GNU Lesser General Public License for more details.\n";


static gboolean show_version;

// legacy commands are frozen; do not extend
const GOptionEntry legacy_opts[] = {
  {"version", 0, 0, G_OPTION_ARG_NONE, &show_version, "Show version", NULL},
  {NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL, NULL}
};

static const struct command root_subcmds[] = {
  {
    .command = &assoc_cmd,
  },
  {
    .command = &prop_cmd,
  },
  {
    .command = &region_cmd,
  },
  {
    .command = &slide_cmd,
  },
  {
    .command = &test_cmd,
  },
  {}
};

static const GOptionEntry root_opts[] = {
  {"version", 0, 0, G_OPTION_ARG_NONE, &show_version, "Show version", NULL},
  {NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL, NULL}
};

static const struct command root_cmd = {
  .description = "Retrieve data from whole slide images.",
  .options = root_opts,
  .subcommands = root_subcmds,
};

static int usage(GOptionContext *octx) {
  g_autofree gchar *help = g_option_context_get_help(octx, true, NULL);
  fprintf(stderr, "%s", help);
  return 2;
}

static const struct command *canonicalize(const struct command *cmd) {
  while (cmd->command) {
    cmd = cmd->command;
  }
  return cmd;
}

static int invoke_cmdline(const struct command *cmd, char *parents,
                          int argc, char **argv) {
  GError *err = NULL;

  g_autofree char *param_str =
    g_strdup_printf("%s%s", parents,
                    cmd->parameter_string ? cmd->parameter_string : "");
  g_autoptr(GOptionContext) octx = g_option_context_new(param_str);
  if (cmd->options) {
    g_option_context_add_main_entries(octx, cmd->options, NULL);
  }
  g_autoptr(GString) summary = g_string_new(cmd->description);
  if (!cmd->description) {
    g_string_append_printf(summary, "%s.", cmd->summary);
  }
  if (cmd->subcommands) {
    g_string_append(summary, "\n\nSubcommands:");
    for (const struct command *subcmd_ = cmd->subcommands;
         canonicalize(subcmd_)->name; subcmd_++) {
      const struct command *subcmd = canonicalize(subcmd_);
      if (subcmd->summary) {
        g_string_append_printf(summary, "\n  %-16s %s",
                               subcmd->name, subcmd->summary);
      }
    }
    // stop at first non-option argument so we can invoke the subcommand
    // handler
    g_option_context_set_strict_posix(octx, true);
  }
  g_option_context_set_summary(octx, summary->str);

  common_parse_options(octx, &argc, &argv, &err);
  if (err) {
    common_warn("%s\n", err->message);
    g_error_free(err);
    return usage(octx);
  }

  if (show_version) {
    fprintf(stderr, version_format, g_get_prgname(), openslide_get_version());
    return 0;
  }

  if (cmd->subcommands) {
    if (argc < 2) {
      return usage(octx);
    }
    for (const struct command *subcmd_ = cmd->subcommands;
         canonicalize(subcmd_)->name; subcmd_++) {
      const struct command *subcmd = canonicalize(subcmd_);
      if (g_str_equal(subcmd->name, argv[1])) {
        g_autofree char *new_parents =
          g_strdup_printf("%s%s ", parents, argv[1]);
        g_free(argv[1]);
        // argc + 1 so we also move trailing NULL
        for (int i = 2; i < argc + 1; i++) {
          argv[i - 1] = argv[i];
        }
        return invoke_cmdline(subcmd, new_parents, argc - 1, argv);
      }
    }
    return usage(octx);
  }

  if (cmd->handler) {
    // Remove "--" arguments; g_option_context_parse() doesn't
    for (int i = 0; i < argc; i++) {
      if (g_str_equal(argv[i], "--")) {
        g_free(argv[i]);
        for (int j = i + 1; j <= argc; j++) {
          argv[j - 1] = argv[j];
        }
        --argc;
        --i;
      }
    }

    // argv[0] becomes argv[-1]
    argc--;
    argv++;
    if (cmd->min_positional && argc < cmd->min_positional) {
      return usage(octx);
    }
    if (cmd->max_positional && argc > cmd->max_positional) {
      return usage(octx);
    }
    return cmd->handler(argc, argv);
  }

  g_assert_not_reached();
}

static char *get_progname(void) {
  const char *prgname = g_get_prgname();
#ifdef _WIN32
  size_t len = strlen(prgname);
  char *ret = g_ascii_strdown(prgname, len);
  if (g_str_has_suffix(ret, ".exe")) {
    ret[len - 4] = 0;
  }
  return ret;
#else
  return g_strdup(prgname);
#endif
}

int main(int argc, char **argv) {
  // properly handle Unicode arguments on Windows; set prgname
  common_fix_argv(&argc, &argv);
  g_autofree char *cmd_name = get_progname();
  const struct command *cmd = &root_cmd;
  if (g_str_equal(cmd_name, "openslide-quickhash1sum")) {
    cmd = &quickhash1sum_cmd;
  } else if (g_str_equal(cmd_name, "openslide-show-properties")) {
    cmd = &show_properties_cmd;
  } else if (g_str_equal(cmd_name, "openslide-write-png")) {
    cmd = &write_png_cmd;
  }
  return invoke_cmdline(cmd, "", argc, argv);
}
